lapack/gonum/dlartg.go

```diff
@@ -9,72 +9,50 @@ import "math"
 // Dlartg generates a plane rotation so that
 //  [ cs sn] * [f] = [r]
 //  [-sn cs]   [g] = [0]
-// This is a more accurate version of BLAS drotg, with the other differences that
-// if g = 0, then cs = 1 and sn = 0, and if f = 0 and g != 0, then cs = 0 and sn = 1.
-// If abs(f) > abs(g), cs will be positive.
+// where cs*cs + sn*sn = 1.
+//
+// This is a more accurate version of BLAS Drotg, with the other differences
+// that
+//  - if g = 0, then cs = 1 and sn = 0
+//  - if f = 0 and g != 0, then cs = 0 and sn = 1
+//  - r takes the sign of f and so cs is always non-negative
 //
 // Dlartg is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlartg(f, g float64) (cs, sn, r float64) {
-	safmn2 := math.Pow(dlamchB, math.Trunc(math.Log(dlamchS/dlamchE)/math.Log(dlamchB)/2))
-	safmx2 := 1 / safmn2
-	if g == 0 {
+	// Implementation based on Supplemental Material to:
+	// Edward Anderson. 2017. Algorithm 978: Safe Scaling in the Level 1 BLAS.
+	// ACM Trans. Math. Softw. 44, 1, Article 12 (July 2017), 28 pages.
+	// DOI: https://doi.org/10.1145/3061665
+	const safmin = dlamchS
+	const safmax = 1 / safmin
+	f1 := math.Abs(f)
+	g1 := math.Abs(g)
+	switch {
+	case g == 0:
 		cs = 1
 		sn = 0
 		r = f
-		return cs, sn, r
-	}
-	if f == 0 {
+	case f == 0:
 		cs = 0
-		sn = 1
-		r = g
-		return cs, sn, r
-	}
-	f1 := f
-	g1 := g
-	scale := math.Max(math.Abs(f1), math.Abs(g1))
-	if scale >= safmx2 {
-		var count int
-		for {
-			count++
-			f1 *= safmn2
-			g1 *= safmn2
-			scale = math.Max(math.Abs(f1), math.Abs(g1))
-			if scale < safmx2 || count == 20 {
-				break
-			}
-		}
-		r = math.Sqrt(f1*f1 + g1*g1)
-		cs = f1 / r
-		sn = g1 / r
-		for i := 0; i < count; i++ {
-			r *= safmx2
-		}
-	} else if scale <= safmn2 {
-		var count int
-		for {
-			count++
-			f1 *= safmx2
-			g1 *= safmx2
-			scale = math.Max(math.Abs(f1), math.Abs(g1))
-			if scale >= safmn2 || count == 20 {
-				break
-			}
-		}
-		r = math.Sqrt(f1*f1 + g1*g1)
-		cs = f1 / r
-		sn = g1 / r
-		for i := 0; i < count; i++ {
-			r *= safmn2
-		}
-	} else {
-		r = math.Sqrt(f1*f1 + g1*g1)
-		cs = f1 / r
-		sn = g1 / r
-	}
-	if math.Abs(f) > math.Abs(g) && cs < 0 {
-		cs *= -1
-		sn *= -1
-		r *= -1
+		sn = math.Copysign(1, g)
+		r = g1
+	case drtmin < f1 && f1 < drtmax && drtmin < g1 && g1 < drtmax:
+		d := math.Sqrt(f*f + g*g)
```

Comment:
`math.Hypot(f, g)`?

Reply:
It would be cleaner but `math.Hypot` does extra work to protect against things that won't happen here, so not calling it can be faster. I guess I'll have to support this with some benchmarks.

Reply:
If the overflow/underflow is not going to happen, then there's no point in the protection. Benchmarks aren't needed.